### PR TITLE
0.4 roadmap

### PR DESCRIPTION
**Pull request by sofroniewn**
# Description
This PR contains a roadmap for our 0.4 series of releases. It also has a look back for our 0.3 series. Right now this is inline with the old 0.3 roadmap, but we could put it at the top of the 0.4 roadmap or in its own document.

This is still very early, but feedback always welcome!

### CODE REVIEW

**Comment by tlambert03** [comment]
I'd like to see a bullet point about refactoring/cleaning existing code.  I don't think any sort of "feature freeze" is in order any time soon, but I do think we should keep revisiting what has been done and refining/clarifying/documenting it... so always good to have that goal explicitly mentioned.

**Comment by sofroniewn** [comment]
> I'd like to see a bullet point about refactoring/cleaning existing code.

Good point - we can see how specific we want to get, but we've got a lot in progress or planned, evented dataclass, color manager, slicing off the layer, asycn clean up, console pullout, more qt / model separation (own model repo?), among others.

Is this the sort of stuff you had in mind @tlambert03? Any of those you want to make sure we mention? @jni already asked for async clean up to go in its corresponding section.

**Comment by tlambert03** [comment]
Yep that kind of stuff!

**Comment by justinelarsen** [comment]
Overall, I think the 0.4 roadmap looks great! I like the inline comments for the look back on the 0.3 series. It helps to provide context and track progress. I also agree with @tlambert03 on calling out the refactoring/cleanup explicitly so it's clear and we make progress on it.

**Comment by sofroniewn** [comment]
Just noting that an in place updatable version of the bundled app was identified as a high priority need

**Comment by sofroniewn** [comment]
Ok @jni this has been updated so 0.3 roadmap lookback is now its own doc and 0.3 roadmap is unchanged

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_3_lookback.md]: I would change the title/subtitle here, e.g.

```suggestion
# Roadmap retrospective

Periodically, the napari team produces roadmaps to clarify priorities in the coming months of development. When a new roadmap is introduced, we will retrospectively examine how well actual development tracked with the appropriate roadmap.

Below, we examine our 0.3.x roadmap (reproduced in its entirety) and compare it to what we achieved in the corresponding 6 months (in inline quote blocks).

## For 0.3.* series of releases - April 2020-Nov 2020

```

Comment (jni) [reply docs/source/developers/ROADMAP_0_3_lookback.md]: I'd also consider moving the initial quote block to its own paragraph above the `## For 0.3`.

I also feel uncomfortable saying stuff like "we've reduced the number of bugs" without numbers. We can find this out using GH search, finding number of issues called "bug" that were closed in that period, and of those, how many were actually there when we created the roadmap — ie how much effort was actually spent on clearing up existing bugs vs creating and fixing new ones. I can try to do this after this review.

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_3_lookback.md]: octree is also now available.

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_3_lookback.md]: Note here that we successfully used perfmon to dramatically improve the performance of the shapes layer

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_3_lookback.md]: a affine -> affine

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_3_lookback.md]: This undersells our progress, I think!

- installation in fresh environments is often successful, especially using conda-forge
- the bundled app exists and works for a large proportion of users
- *but* many platform specific bugs still remain. Windows is often a problem, partly due to lack of representation in the core team, and macOS Big Sur has raised a whole new set of issues, though again, this works with conda-forge and python 3.9.0 🎉

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_3_lookback.md]: Ah great stuff. I'd suggest mentioning this in the intro *or* writing (see below)

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_3_lookback.md]: Again, this feels like we did very little, but actually all of that work is great and works really well! I would rephrase as "we have added programmatic support for text and other properties in points, shapes, and labels, allowing users to effectively visualize many properties in their data. However, support for editing these properties from the UI remains limited or non-existent, so we will aim to improve this in the next roadmap."

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_3_lookback.md]: I don't know how to improve this but it sounds so matter of fact when it is indeed so awesome! 😂

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_3_lookback.md]: remove fairly?

**Review by jni — COMMENTED** [review]
Just added some notes on the retrospective, submitting now but will add further notes for the 0.4 roadmap

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_3_lookback.md]: Worth writing here that we actually have experimental support for this right now!

File: docs/source/developers/ROADMAP_0_4.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_4.md:3]: January 2021?

File: docs/source/developers/ROADMAP_0_4.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_4.md]: The link probably needs updating

File: docs/source/developers/ROADMAP_0_4.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_4.md]: Both of these are now implemented but as opt-in, only programmatically. I would outline two priorities here:
- add persistent settings and use those to allow anyone to use these features, even in the bundled app
- bring the features out of experimental and into mainline napari

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_3_lookback.md]: good points on the organization

on the bug side, we have > 80 when i started tracking over the summer (~July) and then brought that down to < 60 in November. We're now back at 64. Agreed though those comments are best supported with data, which is easy to produce. We can tone it down for the retro, but would be nice to do those sorts of analyses at some point and include figures!

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_3_lookback.md]: will update

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_3_lookback.md]: will add

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_3_lookback.md]: ah it's down here! i forgot i had put that in, great!

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_3_lookback.md]: yes that's fair ;-)

File: docs/source/developers/ROADMAP_0_4.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_4.md]: I don't love this as a priority on its own — it'd be good to mention what this enables in terms of usage and features, e.g. remote viewer, sharing of viewers between different users, deep linking, animation interpolating all attributes of the viewer...

File: docs/source/developers/ROADMAP_0_4.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_4.md]: maybe link to your repo here? Or remove altogether — maybe point to your repo in the retrospective?

File: docs/source/developers/ROADMAP_0_4.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_4.md]: This should be bumped up, and we should mention that it is critical to give users early access to experimental features.

File: docs/source/developers/ROADMAP_0_4.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_4.md]: Already done. Maybe, strengthen support? It would be nice actually to tie it in with the community, e.g. "0.4.3 brought experimental support for functional plugins. During the 0.4.x series, we aim to work with our community to strengthen that support, identify gaps in functionality, and fill those gaps, either with new plugin types or by modifying the existing plugins architecture to suit developer and user needs."

File: docs/source/developers/ROADMAP_0_4.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_4.md]: Delete?

File: docs/source/developers/ROADMAP_0_4.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_4.md]:
```suggestion
- Add accessible **documentation**, tutorials, and demos.
```

File: docs/source/developers/ROADMAP_0_4.md
Comment (jni) [inline docs/source/developers/ROADMAP_0_4.md]:
```suggestion
- Support **functional and interactive** plugins that enable complete image analysis workflows.
```

**Review by jni — APPROVED** [review]
Ok I've added a review for the 0.4 roadmap itself. One high level comment I would add is that each section should have a little introductory paragraph for it before launching into the dot-points. Happy to write those myself.

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_3_lookback.md]: done

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_3_lookback.md]: done

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_3_lookback.md]: added line, don't want to say too much as a feel like it belongs more in our 0.4 section

Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_3_lookback.md]: wasn't really an 0.3 accomplishment, i might actually not include. it will be highlighted elsewhere

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_3_lookback.md]: done

File: docs/source/developers/ROADMAP_0_3_lookback.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_3_lookback.md]: ok more added

File: docs/source/developers/ROADMAP_0_4.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_4.md:3]: well strictly speaking the first 0.4 release was back then and the doc started in november. i think it's ok. we probably shouldn't let it sit as long next time, but i'd probably leave as is now

File: docs/source/developers/ROADMAP_0_4.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_4.md]: yeah, we can cut section

File: docs/source/developers/ROADMAP_0_4.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_4.md]: updated

File: docs/source/developers/ROADMAP_0_4.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_4.md]: can you take this one

File: docs/source/developers/ROADMAP_0_4.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_4.md]: hmm that's fair. I do want to capture this feedback too https://github.com/napari/napari/pull/1906#issuecomment-730779928

Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_4.md]: open to ideas/ suggestions, maybe you can take a pass?

File: docs/source/developers/ROADMAP_0_4.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_4.md]: yes agreed, can you take

File: docs/source/developers/ROADMAP_0_4.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_4.md]: yes very open to that. i'd like some of the perspective of the doc to still be from everything in 0.4, but should be up to date with where we are now. can you take this one

File: docs/source/developers/ROADMAP_0_4.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_4.md]: done

File: docs/source/developers/ROADMAP_0_4.md
Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_4.md]: ok, maybe we cut here and I add link to my repo in retro

Comment (sofroniewn) [reply docs/source/developers/ROADMAP_0_4.md]: done

File: docs/source/developers/ROADMAP_0_4.md
Comment (jni) [reply docs/source/developers/ROADMAP_0_4.md:3]: eh, still feels like backdating a contract, but 🤷. Happy to leave it be.

**Comment by jni** [comment]
@tlambert03 I've written a paragraph about cleanup and refactor in 7cfde24c. Let me know if it captures the gist of what you envisioned it to say. 😊 (cc also @sofroniewn @justinelarsen)

I'm done making changes to the roadmap so further reviews are welcome! Hopefully not much more to be done before merging and cutting the release candidate! 🎉 CC @royerloic too.

**Comment by sofroniewn** [comment]
@jni your edits look great to me! I think it might be a little ambitious to get the asycn & hook_specs out of experimental before 0.5.0, I don't see that as a must, but it's ok to call out that's a goal.

If no one else has any additional comments I'll merge soon!

**Review by tlambert03 — APPROVED** [review]

**Comment by jni** [comment]
> it might be a little ambitious to get the asycn & hook_specs out of experimental before 0.5.0

later... "we are excited to announce napari 0.4.48!" 😜

**Pull request merged**
